plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -90,75 +100,103 @@ class scale_transformer_t : public wf::view_2D
         /* render the transformed view first */
         view_transformer_t::render_with_damage(src_tex, src_box, damage, target_fb);
 
-        /* call all transformers */
-        for (auto& ol : overlays)
+        /* call all overlays */
+        for (auto& p : overlays)
         {
-            (*ol->ol)(target_fb, damage);
+            auto& ol = *(p.second);
+            if (ol.render_hook)
+            {
+                (*ol.render_hook)(target_fb, damage);
+            }
         }
     }
 
-    /**
-     * Add a new overlay that is rendered after this transform.
-     *
-     * @param ol      The callback function that does the rendering.
-     * @param pad     Callback function to calculate extra padding to add.
-     *                Paddings are not cumulative, i.e. if two overlays add a
-     *                top padding, the larger of the two values is used.
-     * @param z_order Relative order; overlays are called in order.
-     */
-    void add_overlay(std::shared_ptr<render_callback_t>& ol, int z_order,
-        std::shared_ptr<padding_callback_t>& pad)
+    /* call pre-render hooks */
+    void call_pre_hooks()
     {
-        auto it =
-            std::find_if(overlays.begin(), overlays.end(),
-                [z_order] (const auto& other)
+        bool view_pad_shrunk  = false;
+        bool view_pad_grown   = false;
+        bool scale_pad_shrunk = false;
+        for (auto& p : overlays)
         {
-            return other->z_order >= z_order;
-        });
+            auto& ol = *(p.second);
+            if (ol.pre_hook)
```

Comment:
Can't we just start with empty pad, extend padding by going through all hooks, then compare with last padding? Seems like it would be simpler and shorter.